pycloudlib/instance.py

```diff
@@ -369,7 +369,26 @@ def _wait_for_cloudinit(self, *, raise_on_failure: bool):
             then this method will raise an `OSError`.
         """
         self._log.debug('_wait_for_cloudinit to complete')
-        has_wait = "--wait" in self.execute("cloud-init status --help").stdout
+
+        result = self.execute("cloud-init status --help")
+
+        if result.failed:
```

Comment:
This has a problem: `cloud-init status --help` on trusty returns non-zero:

```
root@prepared-gelding:~# cloud-init status --help
usage: cloud-init [-h] [--version] [--file FILES] [--debug] [--force]
                  {init,modules,query,single} ...
cloud-init: error: invalid choice: 'status' (choose from 'init', 'modules', 'query', 'single')
root@prepared-gelding:~# echo $?
2
```

So for trusty systems (or any others with a pre-`status` cloud-init), this is always going to wait ~100s and then raise an `OSError`. This is why we were only checking for the text previously: the actual status code is orthogonal to what we're determining.

I think this may indicate a broader problem with this proposal: it's trying to combine a "can we run commands on this system yet" check with a "does cloud-init have status" check, and the two aren't compatible: a non-zero return code is not exceptional for the "status" check, but it _is_ for "can we run commands".

I think you might have more success if you try implementing this check with a command that has a stable exit code across all supported platforms/distros/releases: if `self.execute("whoami")` exits non-zero then it is near certain that that's an issue with the execution mechanism.

(And then, once we've established that the execution mechanism is working, then we can safely call `cloud-init status --help` as we were previously.)

Reply:
Yes, you are completely right here. I missed the trusty issue that you mentioned. I will update the code with your suggestion